src/mlpack/methods/ann/layer/transposed_convolution.hpp

```diff
@@ -0,0 +1,345 @@
+   */
+  template<typename Archive>
+  void serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+  /*
+   * Return the transposed convolution output size.
+   *
+   * @param size The size of the input (row or column).
+   * @param k The size of the filter (width or height).
+   * @param s The stride size (x or y direction).
+   * @param p The size of the padding (width or height).
+   * @return The transposed convolution output size.
+   */
+  size_t TransposedConvOutSize(const size_t size,
+                               const size_t k,
+                               const size_t s,
+                               const size_t p)
+  {
+    size_t out = std::floor(size - k + 2 * p) / s;
+    return out * s + 2 * (k - p) - 1 + ((((size + 2 * p - k) % s) + s) % s);
+  }
+
+  /*
+   * Rotates a 3rd-order tensor counterclockwise by 180 degrees.
+   *
+   * @param input The input data to be rotated.
+   * @param output The rotated output.
+   */
+  template<typename eT>
+  void Rotate180(const arma::Cube<eT>& input, arma::Cube<eT>& output)
+  {
+    output = arma::Cube<eT>(input.n_rows, input.n_cols, input.n_slices);
+
+    // * left-right flip, up-down flip */
+    for (size_t s = 0; s < output.n_slices; s++)
+      output.slice(s) = arma::fliplr(arma::flipud(input.slice(s)));
+  }
+
+  /*
+   * Rotates a dense matrix counterclockwise by 180 degrees.
+   *
+   * @param input The input data to be rotated.
+   * @param output The rotated output.
+   */
+  template<typename eT>
+  void Rotate180(const arma::Mat<eT>& input, arma::Mat<eT>& output)
+  {
+    // * left-right flip, up-down flip */
+    output = arma::fliplr(arma::flipud(input));
+  }
+
+  /*
+   * Pad the given input data.
+   *
+   * @param input The input to be padded.
+   * @param wPad Padding width of the input.
+   * @param hPad Padding height of the input.
+   * @param output The padded output data.
+   */
+  template<typename eT>
+  void Pad(const arma::Mat<eT>& input,
+           size_t wPad,
+           size_t hPad,
+           arma::Mat<eT>& output)
+  {
+    if (output.n_rows != input.n_rows + wPad * 2 ||
+        output.n_cols != input.n_cols + hPad * 2)
+    {
+      output = arma::zeros(input.n_rows + wPad * 2, input.n_cols + hPad * 2);
+    }
+
+    output.submat(wPad, hPad, wPad + input.n_rows - 1,
+        hPad + input.n_cols - 1) = input;
+  }
+
+  /*
+   * Pad the given input data.
+   *
+   * @param input The input to be padded.
+   * @param wPad Padding width of the input.
+   * @param hPad Padding height of the input.
+   * @param output The padded output data.
+   */
+  template<typename eT>
+  void Pad(const arma::Cube<eT>& input,
+           size_t wPad,
+           size_t hPad,
+           arma::Cube<eT>& output)
+  {
+    output = arma::zeros(input.n_rows + wPad * 2,
+        input.n_cols + hPad * 2, input.n_slices);
+
+    for (size_t i = 0; i < input.n_slices; ++i)
+    {
+      Pad<double>(input.slice(i), wPad, hPad, output.slice(i));
```

Comment:
Should this be ``Pad<eT>``?

Reply:
Ah, yes, didn't look much into it. This should be updated in `convolution.hpp` as well.

Reply:
Absolutely, would be great if you could include this in the PR.